src/realm/node_directory.h

```diff
@@ -0,0 +1,105 @@
+#ifndef NODE_DIRECTORY_H
+#define NODE_DIRECTORY_H
+
+#include <atomic>
+#include <cstdint>
+#include <unordered_map>
+#include <vector>
+#include <shared_mutex>
+
+#include "realm/nodeset.h"
+#include "realm/event.h"
+#include "realm/nodeset.h"
+
+namespace Realm {
+  // -----------------------------------------------------------------
+  // Small immutable header describing a peer
+  // ------------------------------------------------------------------
+  struct NodeMeta {
+    NodeID id{NodeID(-1)};
+    uint64_t epoch{0};
+    uint32_t ip{0};
+    uint16_t udp_port{0};
```

Comment:
Do you think if we can make the `NodeMeta` more generic? I mean not tied to UDP. At least not having a `dup_port`, but something like `ControlPlanPort control_port`, such that we can swap the UDP with something else in the future.